### DISCUSSION TITLE
Smart Meter Interface var not found: sml(1 0 300)

### DISCUSSION THREAD

**Discussion by Hootie81** [arendst/Tasmota#11474] [Q&A] [answered]
Hi All,

Im having issues doing the initial setup for the smart meter interface

in the console i get the lines "var not found: sml(1 0 300)" then sml(1 1 "2F3F210D0A")
im guessing that sml isnt compiled into the code properly, i cant figure out what im missing as the script option seems to compile in without issue..

Any help would be appreciated
Cheers
Chris

**Comment by barbudor** [comment, marked as answer]
Hi
You can use [decode-status](https://github.com/tasmota/Tasmota-decode-status) to check that your binary has been biult properly.
If not, share exactly what you've done so we can advise.

**Reply by Hootie81** [reply]
i have tried running the decode status, but it says unable to load script.. im sorry i find all the instructions for tasmota to be very limited and brief 

step by step
open https://gitpod.io#https://github.com/arendst/Tasmota/tree/master 
edit user_config_override.h to include (added just above the final endif statement)
```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
```

rename platformio_override_sample.ini to platformio_override.ini
then file save to save all the edits
run the complile => `platformio run -e tasmota`
it compiles ok 
the resulting firmware.bin is downloaded and is flahed to the d1 mini using esptool
`esptool.py --port COM9 write_flash -fs 1MB -fm dout 0x0 firmware.bin`

i then copy and run the script
```
>D
res=0  
scnt=0    
;For this Example in the >F section  
>F
;count 100ms   
scnt+=1  
switch scnt  
case 6  
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter   
res=sml(1 0 300)  
res=sml(1 1 "2F3F210D0A")  
;1800ms later \> Send ACK and ask for switching to 9600 baud  
case 18  
res=sml(1 1 "063035300D0A")  
;2000ms later \> Switching sml driver to 9600 baud    
case 20  
res=sml(1 0 9600)   
;Restart sequence after 50x100ms    
case 50  
; 5000ms later \> restart sequence    
scnt=0  
ends        
>M 1
+1,12,o,0,9600, ,13
1,0.0.1(@1,Zählernummer,,Meter_number,0  
1,0.9.1(@#),Zeitstempel,Uhr,time-stamp,0  
1,=h===================  
1,1.8.0(@1,HT+NT Zählerstand,KWh,Total_in,3  
1,1.8.1(@1,HT,KWh,HT_Total_in,3  
1,1.8.2(@1,NT,KWh,NT_Total_in,3  
1,=h===================  
1,36.7.0(@1,Power_L1,kW,kW_L1,2  
1,56.7.0(@1,Power_L2,kW,kW_L2,2  
1,76.7.0(@1,Power_L3,kW,kW_L3,2  
1,16.7.0(@1,Σ_L1+L2+L3,kW,kW_L1+L2+L3,2  
1,=h===================  
1,31.7.0(@1,Strom_L1,A,I_L1,2  
1,51.7.0(@1,Strom_L2,A,I_L2,2  
1,71.7.0(@1,Strom_L3,A,I_L3,2  
#
```

**Reply by gemu2015** [reply]
you need to specify
#define USE_SML_SCRIPT_CMD

this is mentioned in both docs but should eventually be added to each example that uses sml(x)

**Reply by Hootie81** [reply]
cant believe I missed that one, Ive been struggling with this for a week.. can confirm this gets rid of the var not found issue.

still cant get it to work though, I can see IR emitting from the IR led however I can see anything from the meter in response

I have tried sensor53 d1 in the console and it accepts it, however there is nothing displayed debug wise.

So trying something a bit more straight forward to start with.. a counter.. I cant get it to work either. I'm using a d1 mini with a hall effect sensor connected to the D4/GPIO2 input. I can set the pin as a button and it works as expected every rotation of the meter pulses the onboard LED and send a button press.. so I know the input is good, however the script below doesn't seem to count up at all
```
>D
>B
->sensor53 r
>M 1
+1,2,c,0,10,GAS
1,1-0:1.8.0\*255(@100,Gas,cbm,Count,3
#
```

**Reply by gemu2015** [reply]
concerning smart meters i can recommend the forum https://forum.creationx.de/forum/
there are several threads on SML driver and also some tutorials.

for the counter, try with pullup and IRQ mode

+1,2,c,1,-10,GAS

**Reply by Hootie81** [reply]
I've tried it that way too without any luck, might have to go back to simple tasmota counters

**Reply by gemu2015** [reply]
your decode line is wrong! it contains a backslash before the * 
did you get that from the examples? if yes it i will correct asap.

1,1-0:1.8.0*255(@100,Gas,cbm,Count,3

**Reply by Hootie81** [reply]
yep cut and paste from the 3 meter combo example

**Reply by Hootie81** [reply]
so ive still been struggling with the power meter so ive reverted to using the simple blinking light on the meter which is working great

I'm trying to get an idea of instantaneous power and have made this little script which works but isnt very neat and doesnt integrate well to home assistant. is there a way to do this calculation and send the value in the same mqtt 

```
>D
;old value
v1=0
;new value
v2=0
;new val - old val * (60 / teleperiod) * 60 
v3=0

>S
; publish inst power every teleperiod time
if mqtts>0
and upsecs%tper==0
then
;save old value
v1=v2
;get new value
v2=sml[2]
;instantaneous power
v3=(v2-v1)*(60/tper)*60
=>Publish tele/%topic%/SENSOR {"Time":"%tstamp%","Power":{"Inst":%v3%}}
endif

>B
->sensor53 r

>M 2
+1,2,c,1,-10,GAS
+2,4,c,1,-10,Power
1,1-0:1.8.0*255(@100,Gas,m3,Count,3
2,1-0:1.8.0*255(@1000,Power,kWh,Count,3
#
```

**Reply by gemu2015** [reply]
did you try with descriptor math ?

e.g. below calculates power delta every 10 seconds

\>M 2
+1,2,c,1,-10,GAS
+2,4,c,1,-10,Power
1,1-0:1.8.0*255(@100,Gas,m3,Count,3
2,1-0:1.8.0*255(@1000,Power,kWh,Count,3
2,=d 2 10 @1,current Power,W,Power_curr,0
#

**Reply by Hootie81** [reply]
@gemu2015  So that messed things up a bit! not sure exactly why, but I noticed the closing curly bracket missing.. so commenting out the line in the script makes it come back. also I noted the extra Power_curr and the value doesn't get transmitted, in fact the MQTT sensor topic stops working all together. its like its missing part of the string to transmit
The value does show up correctly in the web display
note the difference between the 2 lines below, first one with the script exactly as above, and the second is the normal output
```
15:28:22.415 MQT: tele/Utilities/SENSOR = {"Time":"2021-03-30T15:28:22","GAS":{"Count":3.160},"Power":{"Count":78.318}

15:28:42.412 MQT: tele/Utilities/SENSOR = {"Time":"2021-03-30T15:28:42","GAS":{"Count":3.160},"Power":{"Count":78.322}}
```

**Reply by gemu2015** [reply]
\# char at the end of descriptor was missing

**Reply by Hootie81** [reply]
the # is the last line of my script, i just inserted the one extra line in above the #. it was the first thing i checked

**Reply by gemu2015** [reply]
ok tested the descriptor on a local device and it behaves as expected.
(no active inputs, therefore zero values)
17:36:45.812 MQT: tele/sonoff/SENSOR = {"Time":"2021-03-30T17:36:45","GAS":{"Count":12.340},"Power":{"Count":0.000,"Power_curr":0}}

\>M 2
+1,2,c,1,-10,GAS
+2,4,c,1,-10,Power
1,1-0:1.8.0*255(@100,Gas,m3,Count,3
2,1-0:1.8.0*255(@1000,Power,kWh,Count,3
2,=d 2 10 @1,current Power,W,Power_curr,0
\#

???

**Reply by gemu2015** [reply]
using latest dev version

**Reply by Hootie81** [reply]
@gemu2015 first of all thanks very much for you help with the issue. 
im guessing the * in the lines above is removed between the 0 and 255.. 
otherwise i have the same issue missing the Power_curr
16:51:56.431 MQT: tele/Utility/SENSOR = {"Time":"2021-03-30T16:51:56","GAS":{"Count":3.160},"Power":{"Count":79.430}
is there anything else needed at compile for the descriptor math?

**Reply by gemu2015** [reply]
yes \# and \* are special characters and by copy and paste they get lost

no you don't need special defines for that math.

but did you use the latest dev ?  there had been an bug in a previous version with MQTT

**Reply by gemu2015** [reply]
there has been an issue in earlier versions with SML MQTT.  must use latest dev version.

**Reply by Hootie81** [reply]
That was the issue thanks, was thinking the master branch would be more stable so chose that one first.

is there a way to reset the power to 0 if it doesn't receive any pulses within the period? it seems to hold the last value even though the difference is 0.  I noticed once the sun came up the power_curr value became fixed at 700W even though the count didn't change, till the clouds came and I started using power from the grid again.
tested it tonight with the next device for the water meter, just send a couple of pulses then nothing and the count stays static, and so does the rate

>D
>B
->sensor53 r
>M 1
+1,2,c,1,-10,Water
1,1-0:1.8.0*255(@100,Water,m3,Count,3
1,=d 1 10 @1,Flowrate,LPS,water_flow,0
#

**Reply by gemu2015** [reply]
just realized that you are the first one that does math on a single counter only.
then when no more pulses come in the math routines are nor longer triggered.
if you would combine the counter with a meter this would not be the case.

i have to check how that can be solved.

**Reply by gemu2015** [reply]
please check if that solves your issue.
i added a function smld(n) n = number of meter (1 ...n)
this function forces the recalculation of SML math.
you may call it e.g. every 10 seconds or what ever in section >S

[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/6236891/xdrv_10_scripter.ino.zip)

**Reply by Hootie81** [reply]
That works a treat! Thanks so much for your help
I figured the descriptor math must have a timer related to it anyway so to keep the script simple I call the recalculation every second.
I still need to check the conversion to LPM, but seems to be working correctly and dropping to 0. will get my power meter updated and try it out when the suns up tomorrow

```
>D
>S
smld(1)
>B
->sensor53 r
>M 1
+1,2,c,1,-10,Water
1,1-0:1.8.0*255(@100,Water,m3,Count,3
1,=d 1 60 @100,Flowrate,LPM,water_flow,0
#
```
